### PR TITLE
rails 5.1 support

### PR DESCRIPTION
**Pull request by kbrock**
Move our rails 5.1 `virtual_attribute` changes from manageiq to here.

NOTE:

tested:
- `join_keys` change
- `aliases` => `_`

untested:
- `&block`
- `find_with_associations`

see also https://github.com/ManageIQ/manageiq/pull/18076

### CODE REVIEW

**Comment by kbrock** [comment]
@juliancheal and @jrafanie These changes look like the changes we needed to make in the miq repo?

File: Appraisals [hunk 1]
```diff
@@ -3,14 +3,16 @@
   appraise "#{db_gem}-#{ar_version.split('.').first(2).join}" do
     gem "activerecord", "~> #{ar_version}"
 
-    if ar_version >= "5.1"
-      gem "pg"
+    gem "pg" #, " ~> 0.18.4"
```
Comment (NickLaMuro) [inline Appraisals]: Pedantic:  Remove the comment?

**Comment by juliancheal** [comment]
giphy

File: lib/active_record/virtual_attributes/virtual_fields.rb [hunk 2]
```diff
@@ -176,15 +176,22 @@ def without_virtual_includes
...
+        if ActiveRecord.version.to_s >= "5.1"
+          recs, join_dep = real.find_with_associations { |relation, join_dependency| [relation, join_dependency] }
+        else
+          recs = real.find_with_associations
+        end
         MiqPreloader.preload(recs, preload_values + includes_values) if includes_values
 
+        if ActiveRecord.version.to_s >= "5.1"
+          yield recs, join_dep
+        end
```
Comment (jrafanie) [inline lib/active_record/virtual_attributes/virtual_fields.rb]: Can these conditionals be removed?  I thought 5.0 and earlier didn't take advantage of them.  Maybe yield if block_given?  I was hoping we can do the logic for 5.1 in a 5.0 compatible way.

File: lib/active_record/virtual_attributes/virtual_total.rb [hunk 3]
```diff
@@ -110,7 +110,11 @@ def virtual_aggregate_arel(reflection, method_name, column)
 
           foreign_table = reflection.klass.arel_table
           # need db access for the keys, so delaying all this lookup until call time
-          join_keys = reflection.join_keys(reflection.klass)
+          if ActiveRecord.version.to_s >= "5.1"
```
Comment (jrafanie) [inline lib/active_record/virtual_attributes/virtual_total.rb]: I don't know how to do this without the version check.  So I'm 👍 on this.

File: lib/active_record/virtual_attributes/virtual_fields.rb (hunk 2 above)
Comment (kbrock) [reply lib/active_record/virtual_attributes/virtual_fields.rb]: `block_given?` instead?

Comment (kbrock) [reply lib/active_record/virtual_attributes/virtual_fields.rb]: part of the version check tells us how to proceed further when we deprecate 5.0
So we'll know to remove the extra if check.

While version checking is uglier, it does tell us how to deprecate code for older versions

Comment (jrafanie) [reply lib/active_record/virtual_attributes/virtual_fields.rb]: Right, I'm not sure if it's even needed though.  I'm fine if it is.

Comment (jrafanie) [reply lib/active_record/virtual_attributes/virtual_fields.rb]: Looks good 👍

File: lib/active_record/virtual_attributes/virtual_fields.rb [hunk 4]
```diff
@@ -176,15 +176,23 @@ def without_virtual_includes
...
+        else
+          recs = real.find_with_associations
+        end
         MiqPreloader.preload(recs, preload_values + includes_values) if includes_values
 
+        # when 5.0 support is dropped, assume a block given
+        if block_given?
+          yield recs, join_dep
+        end
         recs
```
Comment (jrafanie) [inline lib/active_record/virtual_attributes/virtual_fields.rb]: 🤣 ^  oops

After much debugging @kbrock and I found it should have been:

```ruby
        if block_given?
          yield recs, join_dep
        else
          recs
        end
```

**Pull request merged**